src/writer/libtest.rs

```diff
@@ -0,0 +1,708 @@
+                self.ignored += 1;
+
+                TestEvent::ignored(name)
+            }
+            Step::Failed(_, world, err) => {
+                self.failed += 1;
+
+                TestEvent::failed(name).with_stdout(format!(
+                    "{err}{}",
+                    world.map(|w| format!("\n{:#?}", w)).unwrap_or_default(),
+                ))
+            }
+        }
+        .into();
+
+        vec![ev]
+    }
+
+    /// Generates test case name.
+    fn test_case_name(
+        &mut self,
+        feature: &gherkin::Feature,
+        rule: Option<&gherkin::Rule>,
+        scenario: &gherkin::Scenario,
+        step: Either<event::HookType, (&gherkin::Step, IsBackground)>,
+    ) -> String {
+        let feature = format!(
+            "{} {}",
+            feature.name,
+            feature.path.as_ref().and_then(|p| p.to_str()).map_or_else(
+                || {
+                    self.features_without_path += 1;
+                    self.features_without_path.to_string()
+                },
+                |s| s.escape_default().to_string()
+            ),
+        );
+        let rule = rule
+            .as_ref()
+            .map(|r| format!("{}: {} {}", r.position.line, r.keyword, r.name));
+        let scenario = format!(
+            "{}: {} {}",
+            scenario.position.line, scenario.keyword, scenario.name,
+        );
+        let step = match step {
+            Either::Left(hook) => format!("{hook} hook"),
+            Either::Right((step, is_bg)) => format!(
+                "{}: {}{}{}",
+                step.position.line,
+                is_bg.then_some("Background ").unwrap_or_default(),
+                step.keyword,
+                step.value,
+            ),
+        };
+
+        [Some(feature), rule, Some(scenario), Some(step)]
+            .into_iter()
+            .flatten()
+            .join("::")
+    }
+}
+
+/// Indicator, whether [`Step`] is [`Background`] or not.
+///
+/// [`Background`]: event::Scenario::Background
+/// [`Step`]: gherkin::Step
+type IsBackground = bool;
+
+impl<W: World, O: io::Write> writer::NonTransforming for Libtest<W, O> {}
+
+impl<W: World + Debug, O: io::Write> writer::Failure<W> for Libtest<W, O> {
+    fn failed_steps(&self) -> usize {
+        self.failed
+    }
+
+    fn parsing_errors(&self) -> usize {
+        self.parsing_errors
+    }
+
+    fn hook_errors(&self) -> usize {
+        self.hook_errors
+    }
+}
+
+#[async_trait(?Send)]
+impl<'val, W, Val, Out> Arbitrary<'val, W, Val> for Libtest<W, Out>
+where
+    W: World + Debug,
+    Val: AsRef<str> + 'val,
+    Out: io::Write,
+{
+    async fn write(&mut self, val: Val)
+    where
+        'val: 'async_trait,
+    {
+        self.output
+            .write_line(val.as_ref())
+            .unwrap_or_else(|e| panic!("Failed to write: {e}"));
+    }
+}
+
+/// [`libtest`][1]'s JSON event.
+///
+/// This format isn't stable, so this implementation uses [this PR][1] as a
+/// reference point.
```

Comment:
@tyranron reference to the format is a bit lower, on definition structs

Reply:
@ilslv I saw it, but it's still hard to find a concrete format description there.

Reply:
It's implementation-defined 🙃

Reply:
@ilslv so, refer to the implementation!

Reply:
Yeah, I thought about it, but there are no concrete data structures defining it unfortunately. I don't find [bare `write` functions][1] very helpful, but I guess thats better than nothing. 


[1]: https://github.com/rust-lang/rust/pull/46450/files#diff-7532e9249c821c5504593d02aa99aee1be73466f74bd8b1abc204d5e07e490f4R55-R162